### ISSUE TITLE
Directory name and package name mismatch can cause deletion of classes without recompile

### ISSUE THREAD

**Issue by labkey-susanh** [gradle/gradle#6038]
When there is a mismatch between the package name and the directory name for a Java class, there are no warnings or errors thrown.  The compile will succeed using the package name in the output classes directory structure.  While this is expected behavior from the Java compiler, it can cause an issue with the Gradle incremental compiler.  

The incremental compiler uses the following when determining if there are classes to delete and classes that need to be recompiled:
```
void preparePatterns(Collection<String> staleClasses, PatternSet filesToDelete, PatternSet sourceToCompile) { 
        for (String staleClass : staleClasses) { 
            String path = staleClass.replaceAll("\\.", "/"); 
            filesToDelete.include(path.concat(".class")); 
            filesToDelete.include(path.concat(".java")); 
            filesToDelete.include(path.concat("$*.class")); 
            filesToDelete.include(path.concat("$*.java")); 

            sourceToCompile.include(path.concat(".java")); 
            sourceToCompile.include(path.concat("$*.java")); 
        } 
    } 
```
That is, it converts the package name to a directory name in an attempt to find the source.  If the directory name does not match the package name, it will come up with nothing that needs to be compiled, and so it will end up deleting the stale classes but then not compiling them again.

### Expected Behavior

Ideally, the incremental compiler would be able to find the path to the source code, but if it is going to assume that the path and package names are the same then a warning about the mismatch between directory and package names would be quite helpful.

### Current Behavior

Current behavior is no warning when a mismatch happens, which can result in classes disappearing and then not getting recreated after a change.

### Context

I discovered this when I mistakenly created my classes in a directory with different capitalization than the package name.  When there were changes to the dependencies of one project that caused some classes in my project to become stale, classes were deleted and then not recreated, which caused the jar file to be less than useful.

### Steps to Reproduce (for bugs)

Though easy enough to produce a sample project in which the package name and directory name are different, my attempts thus far to produce staleness with this simple example that will trigger the incremental recompile have been unsuccessful.

[deletingClasses.zip](https://github.com/gradle/gradle/files/2210504/deletingClasses.zip)

### Your Environment

* Gradle 4.7
* Java 8
* Multi-project build

**Comment by baptistemesta** [comment]
I ran also into the problem.
If you make a mistake and you have a folder named e.g. `org.acme` instead of having folder `org` and subfolder `org/acme` it will also delete your classes.

Worst than that, it will push in your local gradle cache your "new" classes with missing files... so even if you do a full clean the build fails. The only way is then to clear the cache.
It was quite hard to debug that!

**Comment by oehme** [comment]
This was fixed in Gradle 6.5, which gets the source-to-class mapping from the compiler API instead of deriving it from the source file name.
